WT-8447 Database corruption from RTS after fast-delete

keithbostic opened the pull request.

keithbostic commented on src/btree/bt_read.c:
```diff
@@ -223,6 +224,8 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
         switch (current_state = ref->state) {
         case WT_REF_DELETED:
+            if (F_ISSET(session, WT_SESSION_ROLLBACK_TO_STABLE))
+                goto read;
```
@kommiharibabu, this change was necessary to keep RTS from skipping deleted pages during the tree walk. I first tried to set one of `WT_READ_DELETED_SKIP` or `WT_READ_DELETED_CHECK` as part of calling the walk code, but the walk code has this:
```
    /* Check whether deleted pages can be skipped. */
    if (!LF_ISSET(WT_READ_DELETED_SKIP))
        LF_SET(WT_READ_DELETED_CHECK);
```
So, if I set `WT_READ_DELETED_SKIP`, the first test in `__wt_page_in_func()` skips deleted pages, and if I don't set `WT_READ_DELETED_SKIP`, the walk code sets `WT_READ_DELETED_CHECK`, and so the second test in `__wt_page_in_func()` skips deleted pages.

The only code that sets `WT_READ_DELETED_SKIP` is `__sync_checkpoint_can_skip()`, so I was hoping you know what's going on here, why there are two flags, and if you could suggest a solution?

kommiharibabu replied:
@keithbostic Apologies for the delayed reply. The WT_READ_DELETED_SKIP flag is set in function `__wt_sync_file` to not to let the checkpoint read any deleted pages. With durable history, we don't read any pages into the memory as part of the checkpoint. This flag is no longer required. How about redesigning this flag and set it by RTS.

keithbostic replied:
Resolving this conversation -- the change @kommiharibabu suggested has been merged into develop and this branch updated to modify it as necessary.

keithbostic commented on src/reconcile/rec_child.c:
```diff
@@ -17,86 +17,53 @@ __rec_child_deleted(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REF *ref, WT_C
+        if (F_ISSET(r, WT_REC_CLEAN_AFTER_REC))
+            return (__wt_set_return(session, EBUSY));
+        *statep = WT_CHILD_ORIGINAL;
+        return (0);
```
@quchenhao, @kommiharibabu, my second question is this change. The code I have in the branch now isn't quite right, it occasionally fails in `test_truncate03.py`, when under load. The failure stack is
```
stack is eviction ->
    reconcile ->
    wt_rec_row_int ->
    wt_rec_child_modify ->
    __wt_page_del_active(session, ref, false) ->
    wt_page_del_active ->
    wt_txn_visible ->
    txn_visible_id:

WT_ASSERT(session, F_ISSET(txn, WT_TXN_HAS_SNAPSHOT) ||
    session->dhandle->checkpoint != NULL);
```

I think the previous version's flag checks:
```
F_ISSET(r, WT_REC_CLEAN_AFTER_REC | WT_REC_VISIBILITY_ERR)
```
prevented this core dump, but I don't off-hand see how that would be the case, and I'm not sure what I should be doing here if eviction calls this code with (or without), acquiring a snapshot.

Can either of you explain this for me?

Thanks!1

quchenhao replied:
If eviction doesn't have a snapshot, we need to compare the txnid with r->last_running like how we do it in __wt_rec_upd_select. I'm not sure how the flags prevented it from crashing before.

kommiharibabu replied:
I don't think the above problem relates to this PR. @AlexeyAnisimovAU reproduced the similar problem on the develop branch. More details are available in WT-8379.

keithbostic replied:
Thanks, @kommiharibabu. There's now a ticket for this one, WT-8604, and I agree, it's not a problem here.

keithbostic commented:
@sauclovian-wt, RTS now checks for prepared transactions in flight and fails, which makes the RTS/fast-truncate test fail. I've fixed it in b54c147, please make sure I haven't broken the test.

keithbostic commented:
For the record, this draft is currently blocked on WT-8604, but I believe it's otherwise correct.

sauclovian-wt commented:
@keithbostic you haven't broken the test, but I'm not sure there's any point in preparing and then committing the transaction before doing the RTS -- might make sense to check that RTS fails in that case, or since that's likely to create assorted headaches, just exclude that combination (prepare + runtime) from the test scenarios.

keithbostic commented:
> @keithbostic you haven't broken the test, but I'm not sure there's any point in preparing and then committing the transaction before doing the RTS -- might make sense to check that RTS fails in that case, or since that's likely to create assorted headaches, just exclude that combination (prepare + runtime) from the test scenarios.

It has the minor effect of testing that RTS rolls

keithbostic commented:
> I'm not sure there's any point in preparing and then committing the transaction before doing the RTS

Fixed in 42d396b96 by discarding that scenario.

keithbostic commented:
@sauclovian-wt, @agorrod, @mickgraham-mongodb, @quchenhao, @kommiharibabu:

I've refreshed the WT-8447 PR (the PR for the fast-truncate/RTS fixes, also described in WT-6676), and added support for old/new releases and upgrade/downgrade. @agorrod and @mickgraham-mongodb are trying to merge this change before 6.0, so there's some pressure for fast turnaround on this one.

There's quite a bit of discussion in WT-6676, and a little in WT-8447, but here's a quick refresh, plus some additional information about upgrade/downgrade.

Reconciliation of row-store internal pages including fast-truncate WT_REFs (that is, WT_REFs of type WT_REF_DELETED), can get into trouble because when they're read back from disk (either in normal processing or by recovery/RTS), we've discarded the WT_PAGE_DELETED timestamp and prepare information from the fast-truncate. The fix in this branch is to write the WT_PAGE_DELETED information as a trailer to the WT_CELL_ADDR_DEL cell type, and then re-creating the WT_PAGE_DELETED structure when reading the page back into memory. The changes to focus on in review are the rewritten `rec_child.c:__rec_child_deleted()` function, and the RTS code for handling WT_REF_DELETED pages, which has been replaced. There's an additional test `test_rollback_to_stable33.py`.

This is a format change for btree internal pages, and we don't know if there are any MongoDB pages in the wild with WT_CELL_ADDR_DEL types on them, so we have to support both old/new databases and eventual upgrade/downgrade paths when we make a release that writes these new format changes in MongoDB builds.

We've added a new #define, WT_STANDALONE_BUILD_FT_FIX, which is set any time WT_STANDALONE_BUILD is configured (just to make it easier to find the affected code). There's a new flag in the btree page header, WT_PAGE_FT_UPDATE, which is set if a cell of type WT_CELL_ADDR_DEL will have the additional fast-truncate timestamp/prepare information. (It's set on all internal pages, regardless of whether or not the page has cells of that type.)

In standalone builds, the new page type with new cell information is always read/written.

For 6.0 MDB builds, we will never write the new page/cell information, but we will read the new cell information for WT_PAGE_FT_UPDATE page types.

### 6.0 MDB Builds

For 6.0 MDB builds, this code only needs to handle old-style pages with the WT_CELL_ADDR_DEL cell type, that is, no flag on the page and no additional information with the WT_CELL_ADDR_DEL cell.

As I said, we have no idea if existing MongoDB builds are writing WT_CELL_ADDR_DEL cells -- but if they are, it's a data consistency problem. This set of changes tries to improve the situation by refusing to write WT_CELL_ADDR_DEL cells during eviction, that is, if we are writing WT_CELL_ADDR_DEL cells, we will stop doing so for eviction and limit such writes to checkpoints.

When reading WT_CELL_ADDR_DEL cells during  normal processing, we will mark the page for eventual deletion (or re-instantiate the page from scratch if a key is subsequently inserted into the page's name space). When reading WT_CELL_ADDR_DEL cells as part of recovery/RTS, the page is also marked for deletion but as it has no associated WT_PAGE_DELETED information, we'll ignore the page entirely. In short, in all cases, the page contents are discarded, regardless of whether or not that's the correct thing to do.

(Just as a review note: I tried to #ifdef the code so the old code would exist side-by-side with the new code, but both RTS and reconciliation got seriously ugly. Instead I ended up trying to recreate the previous behaviors in the new code.)

### Future MDB builds

At some future point, we'll change that, and start writing the new page/cell information in MDB builds. All that should be required is to remove the code ifdef'd on WT_STANDALONE_BUILD_FT_FIX. Upgrade should be identical to standalone builds in this code base.

Downgrade will be trickier, as that will include previously untested code paths. We have to skip over the information appended to WT_CELL_ADDR_DEL cells, but still not set the WT_PAGE_DELETED structures for the re-created WT_REF_DELETED pages, which will then fall into the existing 6.0 MDB build paths where such pages are simply discarded.

quchenhao commented:
> 

Thanks for the implementation. This makes sense to me. Can we force a clean shutdown in downgrade so that all these deleted pages are outdated and we can safely ignore them in the older version?

keithbostic commented:
> Can we force a clean shutdown in downgrade so that all these deleted pages are outdated and we can safely ignore them in the older version?

As far as I know, downgrade requires a clean shutdown first, but I don't think it's enforced by the server.

quchenhao commented on src/btree/bt_page.c:
```diff
@@ -797,22 +797,20 @@ __inmem_row_int(WT_SESSION_IMPL *session, WT_PAGE *page, size_t *sizep)
+                WT_ERR(__wt_calloc_one(session, &ref->ft_info.del));
+                *ref->ft_info.del = unpack.page_del;
+                ref->ft_info.del->previous_state = WT_REF_DISK;
+                ref->ft_info.del->committed = 1;
```
Can we change committed to a bool instead of uint8_t? I know it may impact backward compatibility.

keithbostic replied:
I don't want to touch the `WT_PAGE_DELETED` structure in this change, but we could create another ticket if you think it's worth it.

keithbostic replied:
@quchenhao, I ended up changing WT_PAGE_DELETED, so I fixed this in a1425d7.

quchenhao commented on src/include/btmem.h:
```diff
@@ -77,6 +77,7 @@ struct __wt_page_header {
 #define WT_PAGE_EMPTY_V_NONE 0x04u /* Page has no zero-length values */
 #define WT_PAGE_ENCRYPTED 0x08u    /* Page is encrypted on disk */
 #define WT_PAGE_UNUSED 0x10u       /* Historic lookaside store page updates, no longer used */
+#define WT_PAGE_FT_UPDATE 0x20u    /* Page contains updated fast-truncate information */
```
Alphabetize

keithbostic replied:
I didn't alphabetize this because these bit values are persistent and can never change, that is, they are sorted on bit order rather than name. If I alphabetize them, then the bit values will be out of order.

agorrod replied:
They can't be, because they are written to disk we need to keep the existing values the same.

quchenhao replied:
No need to change the flag value. Just reorder the entries. WT_PAGE_UNUSED actually has the smallest value.

quchenhao commented on src/include/cell_inline.h:
```diff
@@ -189,13 +189,41 @@ __cell_pack_addr_validity(WT_SESSION_IMPL *session, uint8_t **pp, WT_TIME_AGGREG
+    v = 0;
+    if (cell_type == WT_CELL_ADDR_LEAF_NO)
+        FLD_SET(v, WT_CELL_ADDR_DEL_LEAF_NO);
+    if (page_del->prepare_state != 0)
```
There is no macro defined for unprepared state? What about the case the prepared is resolved? In the case we should not treat it differently right?

keithbostic replied:
That's a good catch -- I agree it's likely wrong as is, but I'm not sure what it should be.

If a checkpoint is writing this information, then `prepare_state` could be any of "inprogress", "locked", or "resolved". 

I'm going to need some help, I don't know what is the correct thing to write here. The test to write a proxy child is `__wt_page_del_active(session, ref, true)`, which is always going to return `true` if the prepare isn't resolved.

keithbostic replied:
@quchenhao, I just pushed a change in 6eb1285 that makes the situation better. It asserts the prepare is in-progress as a placeholder, even though that's not correct.

I also cleaned up some field usage that was from a previous version -- I don't think we need to know if this is a leaf, or a leaf without overflow items, but I will stare at it a little more tomorrow.

quchenhao replied:
I don't understand why we are not just checking prepare equals locked or prepared state? Is this because it can be committed concurrently to the check?

keithbostic replied:
We're checkpointing the internal page and we have the child WT_REF locked, and we're looking at the child WT_REF state. I don't know if a prepare can be updated without holding the WT_REF locked or not -- I need some help here. I'm happy to talk through what's happening, but I don't understand how prepare is handled sufficiently to make these changes on my own.

quchenhao commented on src/include/cell_inline.h:
```diff
@@ -205,13 +228,29 @@ __wt_cell_pack_addr(WT_SESSION_IMPL *session, WT_CELL *cell, u_int cell_type, ui
+        WT_ASSERT(session, cell_type == WT_CELL_ADDR_LEAF || cell_type == WT_CELL_ADDR_LEAF_NO);
+        WT_ASSERT(session,
+          page_del->prepare_state == 0 || page_del->prepare_state == WT_PREPARE_INPROGRESS);
+        WT_ASSERT(session, page_del->committed == 1);
```
If it is prepared, why committed is 1?

quchenhao commented on src/txn/txn_rollback_to_stable.c:
```diff
@@ -1224,54 +1224,39 @@ __rollback_abort_updates(WT_SESSION_IMPL *session, WT_REF *ref, wt_timestamp_t r
+    if (ref->state == WT_REF_DELETED) {
+        page_del = ref->ft_info.del;
+        if (page_del == NULL ||
+          (page_del->prepare_state == 0 && page_del->durable_timestamp <= rollback_timestamp))
```
So we never write prepared resolved state to page deleted structure?

keithbostic commented:
I'm giving up for the day -- this isn't getting us anywhere.

I'll try and figure out the prepare code tomorrow and see if I can get something stable. For now, there's no point in reviewing this code.

keithbostic commented:
@quchenhao, I've pushed a set of fixes into the PR and I'd like to have you review them, and once you and I agree, we can queue up the other reviewers.

Here's an explanation of where I think we are.

I think we're OK on everything but prepared transactions: all we care about outside of a prepared transaction is the txnID, timestamp an durable_timetsamp fields, and those are fully written and restored as part of reconciliation and reading pages into memory.

For prepared transactions, we have to additionally deal with the prepare_state field, which is in one of 4 states: WT_PREPARE_INIT, WT_PREPARE_INPROGRESS, WT_PREPARE_LOCKED and WT_PREPARE_RESOLVED.
```
    /*
     * The prepare state is used for transaction prepare to manage visibility and 
     * inheriting prepare state to update_list.
     */
    volatile uint8_t prepare_state;
```

To recap, when we reconcile an internal page with a child WT_REF with  a fast-truncate operation that's either not yet globally visible or which can be rolled back by RTS, we write a "proxy" cell on the parent page. That proxy cell includes the fast-truncate transaction ID plus timestamps and prepared transaction information.

When we read an internal page in from disk that has a fast-truncate proxy cell on  it, we allocate a WT_REF.WT_PAGE_DELETED structure. The txnID and timestamps are  set to their values as of reconciliation. We additionally set the WT_PAGE_DELETED prepare information as follows:
```
    If the child page's fast-truncate prepared state was set:
        WT_PAGE_DELETED.prepare_state = WT_PREPARE_INPROGRESS
    else
        WT_PAGE_DELETED.prepare_state = 0;
```
**WARNING**: the prepared and timestamp values read/written at reconciliation may race with the prepare being resolved underneath a checkpoint of the internal page.

There the two additional fields in the WT_PAGE_DELETED structure. They used to be named "previous_state" and "committed. In this patch I've renamed them to be "previous_ref_state" and "resolved" for clarity. We can ignore both for the purposes of this discussion, and it's obvious to what values they should be set when the proxy cell is read:
```
    /*
     * The previous state of the WT_REF; if the fast-truncate transaction is rolled back 
     * without the page first being instantiated, this is the state to which the WT_REF 
     * returns.
     */
    uint8_t previous_ref_state;

    /*
     * If the fast-truncate transaction has resolved. If we're forced to instantiate a page, 
     * and the resolved flag isn't set, we have to create a list of update structures for 
     * the transaction to resolve in a subsequent abort/commit.
     */
    uint8_t resolved;
```
So, when reading a proxy cell, we always set the created WT_PAGE_DELETED structure as follows:
```
    WT_PAGE_DELETED.previous_ref_state = WT_REF_DISK; /* the page is on disk */
    WT_PAGE_DELETED.resolved = 1 /* there's no running transaction */
```
I think that covers everything except prepared transactions.

First, eviction:

Eviction gets exclusive access to the internal page and checks the child pages. If there's a fast-truncate that's not yet globally visible, eviction fails. However, we may still write a proxy record during eviction if the fast-truncate can be rolled back by RTS. (That is, the difference between a __wt_page_del_active() check for global visibility done when we first check if a page can be evicted, and the additional check of the deletion being before/after the stable timestamp when we decide what to write for the internal page's cell.) 

I think this is also OK. The child page's prepared fast-truncate must have been resolved in order for the internal page eviction to proceed, which means all we care about during RTS is the txnID and timestamps, and we won't want to write
any of the other prepare information. I think we can fix this by changing the __cell_pack_addr_del() function (the one that decides what fast-truncate information to write into the cell, ignores a prepare_state other than WT_PREPARE_LOCKED or WT_PREPARE_INPROGRESS.

I suspect we could relax the eviction test: it seems to me that we could write a proxy cell whether or not the prepared fast-truncate transaction is "active", but at the moment we don't do that.

Second, checkpoints:

We only care about prepared transactions that are visible to us (if they are not visible to us, we ignore them).

Reconciliation of an internal page locks each child WT_REF as part of evaluating  its state for writing, however the prepare_state field can change without locking  the WT_REF (for example, if prepare-transaction is called). So, we must assume that we can see a prepared fast-truncate operation in any state during reconciliation, and further, the state can change from when we first look at it and then decide what to write to the internal page's cell. Further, the timestamps could race between reading them and looking at the prepare state.

The basic structure in this PR works, I believe. We see the prepared_state of the child page's fast-truncate when we write the parent, when we read the parent back in, we don't skip the leaf page, instead we'll read the page in (RTS explicitly reads in deleted pages rather than skipping them), which causes us to instantiate the page with a set of WT_PREPARE_INPROGRESS updates, which are then rolled back.

I worry about the race where we are checkpointing, reconciliation reads a prepared state of WT_PREPARE_INPROGRESS, but the state has switched to resolved by the time we decide to write out the prepared information, or we race while deciding what to write out in the proxy cell. For example, when reading the page, we might assume the prepare was resolved at the time of the checkpoint, and it's not clear to me that is correct.

sueloverso commented:
Not a reviewer, but that detailed explanation seems like it should be in the ticket so that it is preserved.

quchenhao commented on src/txn/txn.c:
```diff
@@ -1663,7 +1663,7 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
             WT_REF_LOCK(session, op->u.ref, &previous_state);
             if (previous_state == WT_REF_DELETED)
-                op->u.ref->ft_info.del->committed = 1;
+                op->u.ref->ft_info.del->resolved = true;
```
I don't understand why if the previous state is WT_REF_DELETED, the fast truncate is resolved.

quchenhao replied:
If it is rolledback, it is not resolved?

keithbostic replied:
> I don't understand why if the previous state is WT_REF_DELETED, the fast truncate is resolved.

If the WT_REF.type hasn't changed from WT_REF_DELETED, we started a fast-truncate and nothing changed, and we committed it, so we mark the WT_REF.del as "resolved".

If the WT_REF.type changed, it means we instantiated the page after the fast-truncate started, and nothing in the WT_PAGE_DELETED structure is useful after commit, so we discard that structure. Because WT_REF.ft_info.del is a union with WT_REF.ft_info.update, freeing the update field frees everything.

The comment above that code has some text, but it may need enhancing.

keithbostic replied:
> If it is rolledback, it is not resolved?

__wt_txn_rollback() calls __wt_delete_page_rollback() which has some similar work:
```
    /*
     * If the page is still "deleted", it's as we left it, simply reset the state. Otherwise, the
     * page is in an in-memory state, which means it was instantiated at some point. Walk any list
     * of update structures and abort them. We can't use the normal read path to get the pages with
     * updates (the original page may have split, so there many be more than one page), because the
     * session may have closed the cursor, we no longer have the reference to the tree required for
     * a hazard pointer. We're safe since pages with unresolved transactions aren't going anywhere.
     */
    if (current_state == WT_REF_DELETED)
        current_state = ref->ft_info.del->previous_ref_state;
    else if ((updp = ref->ft_info.update) != NULL)
        for (; *updp != NULL; ++updp)
            (*updp)->txnid = WT_TXN_ABORTED;

    /*
     * Don't set the WT_PAGE_DELETED transaction ID to aborted, discard any WT_UPDATE list or set
     * the resolved flag to 1; instead, discard the structures, it has the same effect. It's a
     * single call, they're a union of two pointers.
     */
    __wt_free(session, ref->ft_info.del);
```

quchenhao replied:
I see. The name resolved is a bit confusing may be change it back to committed?

keithbostic replied:
Ummmm, sure... :)

keithbostic replied:
Fixed in 664f38e.

quchenhao commented on test/suite/test_rollback_to_stable33.py:
```diff
@@ -0,0 +1,195 @@
+    def test_rollback_to_stable(self):
+        # RTS will fail if there are uncommitted prepared transactions, so skip tests of prepare
+        # with a runtime call to RTS, that doesn't add useful testing scenarios.
+        if self.prepare and not self.crash:
```
The test still fails on prepared?

keithbostic replied:
RTS will fail any time there are active transactions, won't it?

quchenhao replied:
Oh, that's correct.

keithbostic commented:
After discussions with @quchenhao, I've pushed changes in 660394d that hopefully resolve the issue of prepared objects: basically, the code now asserts it never sees (or attempts to write), in-progress prepared fast-truncate operations. This simplifies things by quite a bit and removes some unnecessary backward compatibility code.

quchenhao reviewed (approved):
lgtm

keithbostic commented:
@agorrod, @kommiharibabu, @quchenhao and I are ready for another review pass on WT-8447. Can you please take a close look and let me know if you see problems?

quchenhao commented on src/reconcile/rec_child.c:
```diff
@@ -15,90 +15,80 @@
      */
-    if (ref->addr != NULL && !__wt_page_del_active(session, ref, true)) {
-        WT_RET(__wt_ref_block_free(session, ref));
+    WT_ORDERED_READ(prepare_state, page_del->prepare_state);
```
Should we also add comment here about checkpoint?

keithbostic replied:
The comment immediately above it mentions checkpoint, I think?

quchenhao replied:
Oh, right.

sauclovian-wt commented on src/btree/bt_vrfy_dsk.c:
```diff
@@ -122,6 +122,8 @@ __wt_verify_dsk_image(WT_SESSION_IMPL *session, const char *tag, const WT_PAGE_H
+        LF_CLR(WT_PAGE_FT_UPDATE);
     if (flags != 0)
         WT_RET_VRFY(session, "page at %s has invalid flags set: 0x%" PRIx8, tag, flags);
 
```
Should probably have code in here to crosscheck the fast-delete information. (GH won't let me comment on the right place to put it, which is at line 390.) The timestamp is effectively a global stop time for the entire page, so it can be checked against the aggregate. I think. (Also, if there's still an underlying page, we probably ought to visit it and crosscheck that stop time against all the time windows on it; but that's probably best done as a follow-on ticket.)

keithbostic replied:
Good idea, I agree. Fixed in 1cd9a86, please take a look and confirm it's what you expected.

sauclovian-wt replied:
What you added seems like a good thing to have, but it wasn't what I was thinking of, which was the timestamps. E.g., page_del->timestamp <= page_del->durable_timestamp; but also, ISTM at least, unpack->ta.newest_stop_durable_ts <= page_del->durable_timestamp, unpack->ta.newest_stop_ts <= page_del->timestamp, unpack->ta.newest_stop_txn <= page_del->txnid, and possibly also unpack->ta.prepare == 0 though that's less clear.

(That is: the delete timestamp is a stop time and is not included in the aggregate, so it should be newer than any/all stop times in the aggregate. Plus if there are prepared values on the page, they should have been resolved before the page got deleted, I think, but I'm not absolutely certain that's necessarily true.)

sauclovian-wt replied:
We discussed this briefly and decided to defer these checks to a followup ticket. Then also decided to revert this change as all it really checks is whether the page flag test when unpacking the cell is actually there; if the page contents are wrong (that is, flag set without the extra elements being present, or vice versa) this check won't go off. But since that will desync the page scan from its contents, something else will fail, and that's probably good enough.

sauclovian-wt commented on src/btree/bt_delete.c:
```diff
@@ -193,15 +193,15 @@ __wt_delete_page_rollback(WT_SESSION_IMPL *session, WT_REF *ref)
-     * instead, we discard both structures entirely, it has the same effect. It's a single call,
-     * they're a union of two pointers.
+     * Don't set the WT_PAGE_DELETED transaction ID to aborted, discard any WT_UPDATE list or set
+     * the resolved flag to 1; instead, discard the structures, it has the same effect. It's a
```
(fixed)

keithbostic replied:
Fixed in b74df1e.

keithbostic commented on test/suite/test_rollback_to_stable33.py:
```diff
@@ -0,0 +1,195 @@
+from wtdataset import SimpleDataSet
+from wtscenario import make_scenarios
+
+# test_rollback_to_stableXX.py
```
Fixed in b74df1e.

keithbostic commented on test/suite/test_rollback_to_stable33.py:
```diff
@@ -0,0 +1,195 @@
+
+# test_rollback_to_stableXX.py
+# Test interaction between fast-delete and RTS.
+class test_rollback_to_stableXX(test_rollback_to_stable_base):
```
Fixed in b74df1e.

keithbostic commented on src/include/cell.h:
```diff
@@ -132,20 +132,21 @@
      * 36: 4 timestamps		(uint64_t encoding, max 9 bytes)
      * 18: 2 transaction IDs	(uint64_t encoding, max 9 bytes)
      *  9: associated 64-bit value	(uint64_t encoding, max 9 bytes)
+     * 28: fast-delete information (transaction ID, 2 timestamps, prepare byte)
```
Fixed in 8247123.

keithbostic commented on src/include/cell.h:
```diff
@@ -132,20 +132,21 @@
+     * or even most cases.
      */
-    uint8_t __chunk[1 + 1 + 1 + 7 * WT_INTPACK64_MAXSIZE + WT_INTPACK32_MAXSIZE];
+    uint8_t __chunk[99];
```
Fixed in 8247123.

keithbostic commented on src/include/cell_inline.h:
```diff
@@ -845,6 +866,19 @@ __wt_cell_unpack_safe(WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk, WT_CE
+    if (unpack->raw == WT_CELL_ADDR_DEL && F_ISSET(dsk, WT_PAGE_FT_UPDATE)) {
+        page_del = &unpack_addr->page_del;
+        WT_RET(__wt_vunpack_uint(
+          &p, end == NULL ? 0 : WT_PTRDIFF(end, p), (uint64_t *)&page_del->txnid));
```
Yes, I think that's safe.

I had to dig to understand why transaction IDs are marked volatile, so I added some comments in 9c87cb1.

keithbostic commented on src/btree/bt_page.c:
```diff
@@ -823,6 +819,9 @@ __inmem_row_int(WT_SESSION_IMPL *session, WT_PAGE *page, size_t *sizep)
                 WT_ERR(__wt_page_modify_init(session, page));
                 __wt_page_modify_set(session, page);
```
Fixed in 7beac0a.

There was a similar construct in `__inmem_col_fix()` that I updated as well. We should probably review all uses of `WT_BTREE_READONLY` and setting the tree dirty.

sauclovian-wt replied:
Probably a good idea, yeah.

keithbostic commented on src/reconcile/rec_child.c:
```diff
@@ -15,90 +15,80 @@
-        *statep = WT_CHILD_IGNORE;
+    if (__wt_page_del_active(session, ref, true) ||
+      (conn->txn_global.has_stable_timestamp &&
+        page_del->timestamp >= conn->txn_global.stable_timestamp)) {
```
Nice catch! Fixed in c7de217.

sauclovian-wt commented:
LGTM

agorrod commented on src/include/cell_inline.h:
```diff
@@ -205,13 +205,31 @@ __wt_cell_pack_addr(WT_SESSION_IMPL *session, WT_CELL *cell, u_int cell_type, ui
+#ifdef WT_STANDALONE_BUILD_FT_FIX
+        WT_IGNORE_RET(__wt_vpack_uint(&p, 0, page_del->txnid));
+        WT_IGNORE_RET(__wt_vpack_uint(&p, 0, page_del->timestamp));
+        WT_IGNORE_RET(__wt_vpack_uint(&p, 0, page_del->durable_timestamp));
```
I'm surprised these are included unconditionally - we generally only include time information if it isn't globally visible.

agorrod replied:
Even selectively, I think we should only write the necessary time information.
Why would we include a transaction ID here, if the ID is globally visible when the record is written? It's just wasted space. Or vice versa with the timestamps.

quchenhao replied:
txnid I think is needed for rts to rollback fast truncate committed concurrently to checkpoint.

kommiharibabu replied:
The txnids that are performed in parallel to the checkpoint are not globally visible until the checkpoint is finished, so we don't have the problem of losing the transaction ids.

I do remember that we didn't reset the txnids and timestamps separately due to the issue described in [here](https://jira.mongodb.org/browse/WT-8074?focusedCommentId=4104281&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-4104281).

But I believe we do not have to follow a similar logic for the fast truncates. I agree with Alex that we can reset the transaction ids and timestamps separately for fast truncate operations. But it needs to be commented on clearly why we do this only fast truncate time windows.

keithbostic replied:
@agorrod, I would rather not make this change for a few reasons: (1) as @kommiharibabu noted, we only write this data if the change isn't globally visible, (2) at least some of the values we don't care about are 0s and those are only a single wasted packed byte, (3) I'm nervous about this information being sufficient for future updates to the system (for example, if we were to want to flush in-progress prepared fast-truncate operations some day would the information we need change), and (4) this change has been tricky enough, and we're close enough to a deadline, that I'm not excited about comments explaining why fast-truncate is different from other operations.

If you want this change, someone else will  have to make it, I don't understand the issues well enough.

sauclovian-wt replied:
Also note that in order to know whether the value is there when reading the cell back in, we'd need to add a byte for flags (I think *) and so the maximum possible space savings is one byte, and that only in the case where we need the transaction ID but not the timestamps. To me that seems marginal enough to not bother...

(*) Maybe there's an available flag bit in the cell header, but I think there's a reason we already didn't do that and used a page flag instead.

One thing we can do though that's pretty much free is encode the durable ts as an offset from the main ts.

sauclovian-wt replied:
It looks like we don't encode the durable ts that way elsewhere though so maybe it's not a good idea...

agorrod replied:
ack - thanks for exploring. I'm happy with the resolution.

agorrod commented on src/include/wt_internal.h:
```diff
@@ -377,6 +377,11 @@ typedef uint64_t wt_timestamp_t;
 
+/* Standalone builds include fast-truncate data consistency fixes. */
+#ifdef WT_STANDALONE_BUILD
+#define WT_STANDALONE_BUILD_FT_FIX 1
```
I'd prefer to switch this to a runtime check that is off by default for non-standalone releases. Is that feasible? I think it will make WT-9061 (testing of the backward compatibility part of this) much simpler.
cc @bvpvamsikrishna since I'm handing over that testing effort to you for now.

keithbostic replied:
I agree, this is a good idea. Fixed in a023650.

kommiharibabu commented on src/reconcile/rec_child.c:
```diff
@@ -15,90 +15,80 @@
      */
-    if (ref->addr == NULL) {
-        *statep = WT_CHILD_IGNORE;
+    if (__wt_page_del_active(session, ref, true) ||
```
Not related to this code, but the `__wt_page_del_active` internally uses the `page_del->timestamp` for the visible_all scenario also, it must be `page_del->durable_timestamp`

kommiharibabu replied:
Yes, it is a data inconsistency error. I will create another ticket for it.

kommiharibabu commented on src/reconcile/rec_child.c:
```diff
@@ -15,90 +15,80 @@
-        *statep = WT_CHILD_IGNORE;
+    if (__wt_page_del_active(session, ref, true) ||
+      (conn->txn_global.has_stable_timestamp &&
+        page_del->timestamp > conn->txn_global.stable_timestamp)) {
```
Also, I don't understand the need for this stable timestamp change. When the `__wt_page_del_active` returns false means that the FT durable timestamp is less than the pinned timestamp.

keithbostic replied:
I'm confused by this: I thought the visibility checks were never against the stable TS, that is, we need to check if RTS will potentially undo this committed and visible change, in which case we cannot remove the underlying blocks. Can you please explain further, or make the change you want to see?

sauclovian-wt replied:
I think there are too many `not`s in this to reason about it safely without writing down the intermediate steps.

`__wt_page_del_active` has these cases:
1. There is no page_del structure. That is not true here so we can skip thinking about it.
2. The txnid is `WT_TXN_ABORTED`. I'm not sure that case is handled in this function at this point, also not sure if it comes up. Don't think it's what Hari's talking about though.
3. The prepare state is `INPROGRESS` or `LOCKED`. This is not true here (checked in the previous clause) so we can skip thinking about it.
4. Otherwise it returns `not` the visibility. In order for the stable timestamp check to run, `__wt_page_del_active` must return false, and since we pass `true` for `visible_all` this means that `__wt_txn_visible_all` must return true.

This in turn can happen if:
1. The connection is closing. Not sure what this implies.
2. The timestamp is `WT_TS_NONE`. In this case the additional check is always false so it doesn't add anything.
3. The timestamp is <= the pinned timestamp. In this case because pinned <= oldest and oldest <= stable, the timestamp must be less than stable so the additional check is always false and doesn't add anything.

I think Hari's point is #3. However, after checking this in detail I think we'd better keep the extra check in case the connection is closing. I don't think we want to discard the page just because we're shutting down if the deletion isn't stable yet.

sauclovian-wt replied:
After talking to Keith about this: probably it's better to roll any closing-related logic inside `__wt_page_del_active`.

kommiharibabu replied:
Thanks, @sauclovian-wt for adding more details. Yes, my point is 3 as you mentioned. I got the same feeling about allowing the additional check as we don't want to discard the page during the connection closing. But later I realised that the `WT_CONN_CLOSING` flag was set only when the shutdown checkpoint is finished. 

I am not sure in which scenarios we are going to write that page again after the final checkpoint? Having that check doesn't cause any problems, just want to know any other scenario where we write this page after the final checkpoint?

Yes, I also agree to move it inside the `__wt_page_del_active` would be good.

keithbostic replied:
Thank you, @kommiharibabu, fixed in c751a47.

sauclovian-wt replied:
@kommiharibabu In that case there's nothing to worry about. Which is a relief, because if it weren't true there are other problems the closing check might cause problems.

kommiharibabu replied:
Thanks.

kommiharibabu commented on src/include/cell_inline.h:
```diff
@@ -845,6 +866,19 @@ __wt_cell_unpack_safe(WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk, WT_CE
+    /* Unpack any fast-truncate information. */
+    if (unpack->raw == WT_CELL_ADDR_DEL && F_ISSET(dsk, WT_PAGE_FT_UPDATE)) {
+        page_del = &unpack_addr->page_del;
+        WT_RET(__wt_vunpack_uint(
```
This `page_del->txnid` needs same time window clear logic in `__cell_addr_window_cleanup`. Otherwise, we will read a stale transaction id upon restart.

sauclovian-wt replied:
Good catch; I thought about that but incorrectly decided it wasn't needed because it's a stop time. (The cell-unpacking window cleanup doesn't touch stop times; but it does touch stop txnids.)

kommiharibabu commented on src/include/reconcile_inline.h:
```diff
@@ -371,16 +365,14 @@ __wt_rec_cell_build_addr(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_ADDR *add
     }
 
+    val->cell_len =
+      __wt_cell_pack_addr(session, &val->cell, cell_type, recno, page_del, ta, val->buf.size);
```
Shall we add an assert here to ensure that cell_type is `WT_CELL_ADDR_DEL` when the `page_del` is passed? The main reason behind this is as we access the page_del members if the cell type is DEL in unpack and also we set every internal page with WT_PAGE_FT_UPDATE

keithbostic replied:
In that function, `__wt_rec_cell_build_addr()`, the passed in cell_type will be either `WT_CELL_ADDR_LEAF` or `WT_CELL_ADDR_LEAF_NO`, we switch to the `WT_CELL_ADDR_DEL` cell type inside `__wt_cell_pack_addr()`.

The `__wt_cell_pack_addr()` function does assert the cell type is LEAF or LEAF_NO whenever `page_del` isn't NULL.

Please let me know if I'm misunderstanding your comment.

sauclovian-wt replied:
That reminds me (forgot to mention this yesterday): that assert should be restricted to `WT_CELL_ADDR_LEAF_NO`; we can't (and don't) fast-delete pages with overflow items and the `NO` is what's used to check that further up the call chain.

keithbostic replied:
Thanks, fixed in eb90fd8.

kommiharibabu commented on src/txn/txn_rollback_to_stable.c:
```diff
@@ -1224,54 +1224,40 @@ __rollback_abort_updates(WT_SESSION_IMPL *session, WT_REF *ref, wt_timestamp_t r
+    if (ref->state == WT_REF_DELETED) {
+        page_del = ref->ft_info.del;
+        if (page_del == NULL ||
+          ((page_del->prepare_state == 0 || page_del->prepare_state == WT_PREPARE_RESOLVED) &&
```
Do we need to check for the checkpoint snapshot also here? In case the proxy cell is written by an eviction thread that has a snapshot greater than the checkpoint.

quchenhao replied:
yes, agree

keithbostic replied:
Fixed in 053e357.

Would you please review this change again? -- Thanks.

kommiharibabu reviewed (commented):
The overall logic of writing the FT information is good. A couple of comments.

sauclovian-wt reviewed (approved):
Updates LGTM

quchenhao reviewed (approved):
lgtm

kommiharibabu reviewed (approved):
LGTM

keithbostic commented:
Full WT patch build: https://evergreen.mongodb.com/version/62536c8e850e613a377ebb9b?redirect_spruce_users=true

agorrod commented:
Sadly - the PR changes here look kind of messed up now. They seem to include the checkpoint changes that have been reverted?

agorrod commented:
I'm good to go with this change.

agorrod commented:
The PR looks OK to me now - I'm not sure what happened earlier? Sorry for the misdirect.

kommiharibabu reviewed (approved):
LGTM

keithbostic commented:
Test run with ASAN and stress jobs here: https://spruce.mongodb.com/version/62585e8d7742ae0e72c282c7/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC
